torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -933,21 +935,56 @@ def _run_param_init_fn():
         self._hook_registered = False
 
     def _init_param_exec_order_wrap_policy(self, *args, **kwargs) -> None:
+        auto_wrap_policy = kwargs["auto_wrap_policy"]
+        module = kwargs["module"]
+        if isinstance(
+            auto_wrap_policy.tracing_config,
+            TracingConfig
+        ):
+            tracer = auto_wrap_policy.tracing_config.tracer
+            execution_info = _ExecutionInfo(module)
+            with _patch_tracer(
+                tracer=tracer,
+                root_module=module,
+                execution_info=execution_info,
+            ):
+                tracer.trace(module, auto_wrap_policy.tracing_config.concrete_args)
         # The initial FSDP wrapping is done with auto_wrap_policy.init_policy
+        kwargs["auto_wrap_policy"] = auto_wrap_policy.init_policy
         self.__init__(*args, **kwargs)
```

Comment:
This is prior to this PR but looks a bit weird. IIUC, we have `__init__` -> `_init_param_exec_order_wrap_policy` -> `__init__`, and depending on the type of `kwargs["auto_wrap_policy"]`, we will walk different paths in `__init__`.

@zhaojuanmao @rohan-ahluwalia @awgu could you please help me understand the design reasoning behind this recursive call? Thanks!

Reply:
@mrshenli sorry about the confusion! Originally, the plan for non-recursive policy is as follows:
1. based on the `auto_wrap_policy.init_policy`, recursively wrap the input model (that's why you see we call `__init__` back), and run the FSDP model with the recursive wrap for one iteration to get the parameter/module execution order.
2. starting from the second iteration, remove the FSDP recursive wraps and run non-recursive algorithms.

The step 1 is a bit awkward since we go back to the recursive policy, and this will be removed soon. I and @awgu will soon implement an `_init_param_exec_order_wrap_policy` such that we don't need to go back to the recursive policy, and the new implementation will be on top of #79652.

Reply:
Got it, thanks for the explanation!